File: Sources/WhisperDictationCore/AppConfig.swift
import Foundation

public struct AppConfig: Codable, Sendable {
    public static let defaultPersistRecentCaptures = false
    public static let defaultServerRequestTimeoutSeconds = 30.0
    public static let defaultRobustServerRequestTimeoutSeconds = 120.0
    public static let defaultCLITimeoutSeconds = 90.0
    public static let defaultRobustWhisperServerPort = 8178
    public static let defaultWarmRobustServerOnLaunch = false
    public static let allowedControlHosts = ["127.0.0.1", "localhost", "::1"]

    public let controlHost: String
    public let controlPort: Int
    public let preferredInputDevice: String?
    public let enforcePreferredInputDevice: Bool
    public let prebufferMilliseconds: Int
    public let audioBufferSizeFrames: Int
    public let pollIntervalMilliseconds: Int
    public let whisperServerBinary: String
    public let whisperCliBinary: String
    public let whisperModelPath: String
    public let whisperVADModelPath: String?
    public let whisperServerHost: String
    public let whisperServerPort: Int
    public let whisperRobustModelPath: String?
    public let robustWhisperServerPort: Int
    public let tempDirectory: String
    public let salvageDirectory: String
    public let daemonLogPath: String
    public let whisperServerLogPath: String
    public let controlBinaryPath: String
    public let daemonBinaryPath: String
    public let warmServerOnLaunch: Bool
    public let warmRobustServerOnLaunch: Bool
    public let whisperThreads: Int
    public let persistRecentCaptures: Bool
    public let serverRequestTimeoutSeconds: Double
    public let robustServerRequestTimeoutSeconds: Double
    public let cliTimeoutSeconds: Double

    private enum CodingKeys: String, CodingKey {
        case controlHost
        case controlPort
        case preferredInputDevice
        case enforcePreferredInputDevice
        case prebufferMilliseconds
        case audioBufferSizeFrames
        case pollIntervalMilliseconds
        case whisperServerBinary
        case whisperCliBinary
        case whisperModelPath
        case whisperVADModelPath
        case whisperServerHost
        case whisperServerPort
        case whisperRobustModelPath
        case robustWhisperServerPort
        case tempDirectory
        case salvageDirectory
        case daemonLogPath
        case whisperServerLogPath
        case controlBinaryPath
        case daemonBinaryPath
        case warmServerOnLaunch
        case warmRobustServerOnLaunch
        case whisperThreads
        case persistRecentCaptures
        case serverRequestTimeoutSeconds
        case robustServerRequestTimeoutSeconds
        case cliTimeoutSeconds
    }

    public static let defaultConfigPath =
        (NSHomeDirectory() as NSString).appendingPathComponent(
            "Library/Application Support/WhisperDictation/config.json"
        )

    public static func load(from path: String = AppConfig.defaultConfigPath) throws -> AppConfig {
        let url = URL(fileURLWithPath: path)
        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode(AppConfig.self, from: data)
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        controlHost = try container.decode(String.self, forKey: .controlHost)
        guard Self.isLoopbackControlHost(controlHost) else {
            throw DecodingError.dataCorruptedError(
                forKey: .controlHost,
                in: container,
                debugDescription: "controlHost must be loopback-only: \(Self.allowedControlHosts.joined(separator: ", "))"
            )
        }

        controlPort = try container.decode(Int.self, forKey: .controlPort)
        preferredInputDevice = try container.decodeIfPresent(String.self, forKey: .preferredInputDevice)
        enforcePreferredInputDevice = try container.decode(Bool.self, forKey: .enforcePreferredInputDevice)
        prebufferMilliseconds = try container.decode(Int.self, forKey: .prebufferMilliseconds)
        audioBufferSizeFrames = try container.decode(Int.self, forKey: .audioBufferSizeFrames)
        pollIntervalMilliseconds = try container.decode(Int.self, forKey: .pollIntervalMilliseconds)
        whisperServerBinary = try container.decode(String.self, forKey: .whisperServerBinary)
        whisperCliBinary = try container.decode(String.self, forKey: .whisperCliBinary)
        whisperModelPath = try container.decode(String.self, forKey: .whisperModelPath)
        whisperVADModelPath = try container.decodeIfPresent(String.self, forKey: .whisperVADModelPath)
        whisperServerHost = try container.decode(String.self, forKey: .whisperServerHost)
        whisperServerPort = try container.decode(Int.self, forKey: .whisperServerPort)
        whisperRobustModelPath = try container.decodeIfPresent(String.self, forKey: .whisperRobustModelPath)
        robustWhisperServerPort = try container.decodeIfPresent(Int.self, forKey: .robustWhisperServerPort)
            ?? Self.defaultRobustWhisperServerPort
        tempDirectory = try container.decode(String.self, forKey: .tempDirectory)
        salvageDirectory = try container.decode(String.self, forKey: .salvageDirectory)
        daemonLogPath = try container.decode(String.self, forKey: .daemonLogPath)
        whisperServerLogPath = try container.decode(String.self, forKey: .whisperServerLogPath)
        controlBinaryPath = try container.decode(String.self, forKey: .controlBinaryPath)
        daemonBinaryPath = try container.decode(String.self, forKey: .daemonBinaryPath)
        warmServerOnLaunch = try container.decode(Bool.self, forKey: .warmServerOnLaunch)
        warmRobustServerOnLaunch = try container.decodeIfPresent(Bool.self, forKey: .warmRobustServerOnLaunch)
            ?? Self.defaultWarmRobustServerOnLaunch
        whisperThreads = try container.decode(Int.self, forKey: .whisperThreads)
        persistRecentCaptures = try container.decodeIfPresent(Bool.self, forKey: .persistRecentCaptures)
            ?? Self.defaultPersistRecentCaptures
        let serverRequestTimeoutSeconds = try container.decodeIfPresent(Double.self, forKey: .serverRequestTimeoutSeconds)
            ?? Self.defaultServerRequestTimeoutSeconds
        let robustServerRequestTimeoutSeconds =
            try container.decodeIfPresent(Double.self, forKey: .robustServerRequestTimeoutSeconds)
            ?? Self.defaultRobustServerRequestTimeoutSeconds
        let cliTimeoutSeconds = try container.decodeIfPresent(Double.self, forKey: .cliTimeoutSeconds)
            ?? Self.defaultCLITimeoutSeconds
        guard serverRequestTimeoutSeconds > 0 else {
            throw DecodingError.dataCorruptedError(
                forKey: .serverRequestTimeoutSeconds,
                in: container,
                debugDescription: "serverRequestTimeoutSeconds must be positive"
            )
        }
        guard robustServerRequestTimeoutSeconds > 0 else {
            throw DecodingError.dataCorruptedError(
                forKey: .robustServerRequestTimeoutSeconds,
                in: container,
                debugDescription: "robustServerRequestTimeoutSeconds must be positive"
            )
        }
        guard cliTimeoutSeconds > 0 else {
            throw DecodingError.dataCorruptedError(
                forKey: .cliTimeoutSeconds,
                in: container,
                debugDescription: "cliTimeoutSeconds must be positive"
            )
        }
        self.serverRequestTimeoutSeconds = serverRequestTimeoutSeconds
        self.robustServerRequestTimeoutSeconds = robustServerRequestTimeoutSeconds
        self.cliTimeoutSeconds = cliTimeoutSeconds
    }

    public static func isLoopbackControlHost(_ host: String) -> Bool {
        allowedControlHosts.contains(host.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())
    }
}

public struct AppPaths {
    public let tempDirectoryURL: URL
    public let salvageDirectoryURL: URL
    public let daemonLogURL: URL
    public let whisperServerLogURL: URL
    public let robustWhisperServerLogURL: URL

    public init(config: AppConfig) {
        tempDirectoryURL = URL(fileURLWithPath: config.tempDirectory, isDirectory: true)
        salvageDirectoryURL = URL(fileURLWithPath: config.salvageDirectory, isDirectory: true)
        daemonLogURL = URL(fileURLWithPath: config.daemonLogPath, isDirectory: false)
        whisperServerLogURL = URL(fileURLWithPath: config.whisperServerLogPath, isDirectory: false)
        let robustBase = whisperServerLogURL.deletingPathExtension().lastPathComponent + "-robust"
        if whisperServerLogURL.pathExtension.isEmpty {
            robustWhisperServerLogURL = whisperServerLogURL
                .deletingLastPathComponent()
                .appendingPathComponent(robustBase, isDirectory: false)
        } else {
            robustWhisperServerLogURL = whisperServerLogURL
                .deletingLastPathComponent()
                .appendingPathComponent(
                    "\(robustBase).\(whisperServerLogURL.pathExtension)",
                    isDirectory: false
                )
        }
    }

    public func ensureDirectoriesExist() throws {
        let fm = FileManager.default
        try fm.createDirectory(at: tempDirectoryURL, withIntermediateDirectories: true)
        try fm.createDirectory(at: salvageDirectoryURL, withIntermediateDirectories: true)
        try fm.createDirectory(at: daemonLogURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        try fm.createDirectory(at: whisperServerLogURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        try fm.createDirectory(at: robustWhisperServerLogURL.deletingLastPathComponent(), withIntermediateDirectories: true)
    }
}

File: app/src/core/appConfig.ts
export const DEFAULT_PERSIST_RECENT_CAPTURES = false;
export const DEFAULT_PERSIST_HISTORY = true;
export const DEFAULT_SERVER_REQUEST_TIMEOUT_SECONDS = 30;
export const DEFAULT_ROBUST_SERVER_REQUEST_TIMEOUT_SECONDS = 120;
export const DEFAULT_CLI_TIMEOUT_SECONDS = 90;
export const DEFAULT_ROBUST_WHISPER_SERVER_PORT = 8_178;
export const DEFAULT_WARM_ROBUST_SERVER_ON_LAUNCH = false;
export const DEFAULT_MAC_CAPTURE_BACKEND = "native" as const;
export const ALLOWED_CONTROL_HOSTS = ["127.0.0.1", "localhost", "::1"] as const;

export type MacCaptureBackend = "native" | "electron";

export interface AppConfig {
  controlHost: string;
  controlPort: number;
  preferredInputDevice: string | null;
  enforcePreferredInputDevice: boolean;
  macCaptureBackend?: MacCaptureBackend;
  prebufferMilliseconds: number;
  audioBufferSizeFrames: number;
  pollIntervalMilliseconds: number;
  whisperServerBinary: string;
  whisperCliBinary: string;
  whisperModelPath: string;
  whisperVADModelPath: string | null;
  whisperServerHost: string;
  whisperServerPort: number;
  whisperRobustModelPath: string | null;
  robustWhisperServerPort: number;
  tempDirectory: string;
  salvageDirectory: string;
  daemonLogPath: string;
  whisperServerLogPath: string;
  controlBinaryPath: string;
  daemonBinaryPath: string;
  warmServerOnLaunch: boolean;
  warmRobustServerOnLaunch: boolean;
  whisperThreads: number;
  persistRecentCaptures: boolean;
  persistHistory: boolean;
  serverRequestTimeoutSeconds: number;
  robustServerRequestTimeoutSeconds: number;
  cliTimeoutSeconds: number;
}

export interface AppPaths {
  tempDirectory: string;
  salvageDirectory: string;
  daemonLogPath: string;
  whisperServerLogPath: string;
  robustWhisperServerLogPath: string;
}

export function isLoopbackControlHost(host: string): boolean {
  const normalized = host.trim().toLowerCase();
  return ALLOWED_CONTROL_HOSTS.some((allowed) => allowed === normalized);
}

export function parseAppConfig(value: unknown): AppConfig {
  const source = requireObject(value, "config");
  const controlHost = requireString(source, "controlHost");
  if (!isLoopbackControlHost(controlHost)) {
    throw new Error(
      `controlHost must be loopback-only: ${ALLOWED_CONTROL_HOSTS.join(", ")}`,
    );
  }

  const serverRequestTimeoutSeconds = optionalNumber(
    source,
    "serverRequestTimeoutSeconds",
    DEFAULT_SERVER_REQUEST_TIMEOUT_SECONDS,
  );
  const robustServerRequestTimeoutSeconds = optionalNumber(
    source,
    "robustServerRequestTimeoutSeconds",
    DEFAULT_ROBUST_SERVER_REQUEST_TIMEOUT_SECONDS,
  );
  const cliTimeoutSeconds = optionalNumber(
    source,
    "cliTimeoutSeconds",
    DEFAULT_CLI_TIMEOUT_SECONDS,
  );
  requirePositive(serverRequestTimeoutSeconds, "serverRequestTimeoutSeconds");
  requirePositive(
    robustServerRequestTimeoutSeconds,
    "robustServerRequestTimeoutSeconds",
  );
  requirePositive(cliTimeoutSeconds, "cliTimeoutSeconds");

  return {
    controlHost,
    controlPort: requireInteger(source, "controlPort"),
    preferredInputDevice: optionalNullableString(source, "preferredInputDevice"),
    enforcePreferredInputDevice: requireBoolean(
      source,
      "enforcePreferredInputDevice",
    ),
    macCaptureBackend: optionalMacCaptureBackend(source),
    prebufferMilliseconds: requireInteger(source, "prebufferMilliseconds"),
    audioBufferSizeFrames: requireInteger(source, "audioBufferSizeFrames"),
    pollIntervalMilliseconds: requireInteger(source, "pollIntervalMilliseconds"),
    whisperServerBinary: requireString(source, "whisperServerBinary"),
    whisperCliBinary: requireString(source, "whisperCliBinary"),
    whisperModelPath: requireString(source, "whisperModelPath"),
    whisperVADModelPath: optionalNullableString(source, "whisperVADModelPath"),
    whisperServerHost: requireString(source, "whisperServerHost"),
    whisperServerPort: requireInteger(source, "whisperServerPort"),
    whisperRobustModelPath: optionalNullableString(
      source,
      "whisperRobustModelPath",
    ),
    robustWhisperServerPort: optionalInteger(
      source,
      "robustWhisperServerPort",
      DEFAULT_ROBUST_WHISPER_SERVER_PORT,
    ),
    tempDirectory: requireString(source, "tempDirectory"),
    salvageDirectory: requireString(source, "salvageDirectory"),
    daemonLogPath: requireString(source, "daemonLogPath"),
    whisperServerLogPath: requireString(source, "whisperServerLogPath"),
    controlBinaryPath: requireString(source, "controlBinaryPath"),
    daemonBinaryPath: requireString(source, "daemonBinaryPath"),
    warmServerOnLaunch: requireBoolean(source, "warmServerOnLaunch"),
    warmRobustServerOnLaunch: optionalBoolean(
      source,
      "warmRobustServerOnLaunch",
      DEFAULT_WARM_ROBUST_SERVER_ON_LAUNCH,
    ),
    whisperThreads: requireInteger(source, "whisperThreads"),
    persistRecentCaptures: optionalBoolean(
      source,
      "persistRecentCaptures",
      DEFAULT_PERSIST_RECENT_CAPTURES,
    ),
    persistHistory: optionalBoolean(
      source,
      "persistHistory",
      DEFAULT_PERSIST_HISTORY,
    ),
    serverRequestTimeoutSeconds,
    robustServerRequestTimeoutSeconds,
    cliTimeoutSeconds,
  };
}

function optionalMacCaptureBackend(source: JsonObject): MacCaptureBackend {
  const value = source.macCaptureBackend;
  if (value === undefined || value === null) {
    return DEFAULT_MAC_CAPTURE_BACKEND;
  }
  if (value !== "native" && value !== "electron") {
    throw new Error("macCaptureBackend must be native or electron");
  }
  return value;
}

export function appPaths(config: AppConfig): AppPaths {
  return {
    tempDirectory: config.tempDirectory,
    salvageDirectory: config.salvageDirectory,
    daemonLogPath: config.daemonLogPath,
    whisperServerLogPath: config.whisperServerLogPath,
    robustWhisperServerLogPath: robustLogPath(config.whisperServerLogPath),
  };
}

function robustLogPath(path: string): string {
  const separatorIndex = Math.max(path.lastIndexOf("/"), path.lastIndexOf("\\"));
  const directory = path.slice(0, separatorIndex + 1);
  const filename = path.slice(separatorIndex + 1);
  const dotIndex = filename.lastIndexOf(".");
  if (dotIndex <= 0) {
    return `${directory}${filename}-robust`;
  }
  return `${directory}${filename.slice(0, dotIndex)}-robust${filename.slice(dotIndex)}`;
}

type JsonObject = Record<string, unknown>;

function requireObject(value: unknown, field: string): JsonObject {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${field} must be an object`);
  }
  return value as JsonObject;
}

function requireString(source: JsonObject, field: string): string {
  const value = source[field];
  if (typeof value !== "string") {
    throw new Error(`${field} must be a string`);
  }
  return value;
}

function optionalNullableString(source: JsonObject, field: string): string | null {
  const value = source[field];
  if (value === undefined || value === null) {
    return null;
  }
  if (typeof value !== "string") {
    throw new Error(`${field} must be a string or null`);
  }
  return value;
}

function requireNumber(source: JsonObject, field: string): number {
  const value = source[field];
  if (typeof value !== "number") {
    throw new Error(`${field} must be a number`);
  }
  return value;
}

function requireInteger(source: JsonObject, field: string): number {
  const value = requireNumber(source, field);
  if (!Number.isSafeInteger(value)) {
    throw new Error(`${field} must be an integer`);
  }
  return value;
}

function optionalNumber(
  source: JsonObject,
  field: string,
  defaultValue: number,
): number {
  const value = source[field];
  if (value === undefined || value === null) {
    return defaultValue;
  }
  if (typeof value !== "number") {
    throw new Error(`${field} must be a number`);
  }
  return value;
}

function optionalInteger(
  source: JsonObject,
  field: string,
  defaultValue: number,
): number {
  const value = source[field];
  if (value === undefined || value === null) {
    return defaultValue;
  }
  return requireInteger(source, field);
}

function requireBoolean(source: JsonObject, field: string): boolean {
  const value = source[field];
  if (typeof value !== "boolean") {
    throw new Error(`${field} must be a boolean`);
  }
  return value;
}

function optionalBoolean(
  source: JsonObject,
  field: string,
  defaultValue: boolean,
): boolean {
  const value = source[field];
  if (value === undefined || value === null) {
    return defaultValue;
  }
  if (typeof value !== "boolean") {
    throw new Error(`${field} must be a boolean`);
  }
  return value;
}

function requirePositive(value: number, field: string): void {
  if (value <= 0) {
    throw new Error(`${field} must be positive`);
  }
}
